### PR TITLE
Expose macro `Call` context via new `@caller` macro ivar

### LINKED ISSUE #14055: Exposing a macro's caller
I'm proposing to give macros access to its caller.

This would be benefitial for several use cases, for example:

* Incorporate doc comment attached to the macro call (#14039)
* Raise macro errors or warnings at the location of the call in user code instead of pointing into library code.

I'm sure there can be other useful applications once this feature is available.

A syntax proposal would be a macro instance method named `@caller` whose value is the `Call` instance which calls the macro.
An implementation of this proposal is available in #14048.

```cr
macro foo
  {{ @caller.line_number }}
end

foo # => 5
```

Theoretically, this mechanism could even provide a chain for nested macro calls. I'm not sure if that would be super useful for anything. But I could see it might be handy for deeply nested macros in library code. However, they could probably also pass references to the entry point callers downwards. So it might not be super important.
We should at least consider making `@caller` an array, though.

Another question is whether there might be a need to access a caller's context, particularly the `@type` and `@def` variables.
Perhaps those could be made generally available on `ASTNode` instances, as ivar reads? i.e. `@caller.@type` would refer to the caller's type scope.
This doesn't need to be part of an initial implementation as I'm not sure about real use cases, but it's good to consider optional enhancements like this in the initial feature discussion.

For reference, a similar mechanism for regular methods is the [`::caller` method](https://crystal-lang.org/api/1.10.1/toplevel.html#caller:Array(String)-class-method) which provides a stack trace at runtime. Unlike macros, method instantiations usually cannot be traced back to a specific call at compile time because they are shared. The actual caller is only available at runtime.

### PR DESCRIPTION
**Pull request by Blacksmoke16**
resolves #14055

### CODE REVIEW

**Comment by straight-shoota** [comment]
This is a major new API feature and I believe it actually needs a bit more dedicated discussion first. Let's to that in #14055.

**Comment by Blacksmoke16** [comment]
Marking this as ready for review again. This should be a good initial implementation based on the discussion in #14055.

File: spec/compiler/codegen/macro_spec.cr [hunk 1]
```diff
@@ -1895,4 +1895,34 @@ describe "Code gen: macro" do
...
+
+  it "returns an array of each call" do
+    run(%(
+      macro test
+        {{@caller.size == 1 ? 1 : 0}}
+      end
+
+      test
+      )).to_i.should eq(1)
+  end
```
Comment (straight-shoota) [inline spec/compiler/codegen/macro_spec.cr:1907]: **suggestion:** Move these specs out of `codegen` and refactor with `assert_type`.
The feature should be completely semantical, so there's no need to actually compile any code, just typing it should be sufficient.
For example:
```suggestion
  it "returns an array of each call" do
    assert_type(<<-CRYSTAL) { int32 }
      macro test
        {{ @caller.size == 1 ? 1 : 'a' }}
      end

      test
      CRYSTAL
  end
```
This applies to all specs here. I suppose they could go into `semantic/macro_spec.cr`.

File: spec/compiler/codegen/macro_spec.cr [hunk 2]
```diff
@@ -1895,4 +1895,34 @@ describe "Code gen: macro" do
...
+
+  it "provides access to the macro call stack via @caller" do
+    run(%(
+      macro test(num)
+        {{@caller.last.args[0] == 1 ? 1 : 0}}
```
Comment (straight-shoota) [inline spec/compiler/codegen/macro_spec.cr:1912]: **nitpick:** This should be `first`, not `last`.
If we add more stack frames to `@caller`, they would be added to the end, i.e. the closest one is always the first (same as with `caller`).

```suggestion
        {{@caller.first.args[0] == 1 ? 1 : 0}}
```

Comment (Blacksmoke16) [reply spec/compiler/codegen/macro_spec.cr:1912]: Ah thanks, was thinking in terms of the trace where the last frame is the one at the end. I'll make a note to include this in the book PR for this ivar.

**Review by straight-shoota — APPROVED** [review]
This should be good for a first implementation.

**Pull request merged**